subsystem.go

```diff
@@ -0,0 +1,93 @@
+package tflog
+
+import (
+	"context"
+
+	"github.com/hashicorp/go-hclog"
+)
+
+const newSubsystemLoggerWarning = "This log was generated by a subsystem logger that wasn't created before being used. Use tflog.NewSubsystem to create this logger before it is used."
+
+func getSubsystemLogger(ctx context.Context, subsystem string) hclog.Logger {
+	logger := ctx.Value(providerSpaceRootLoggerKey + loggerKey("."+subsystem))
+	if logger == nil {
+		return nil
+	}
+	return logger.(hclog.Logger)
+}
+
+func setSubsystemLogger(ctx context.Context, subsystem string, logger hclog.Logger) context.Context {
+	return context.WithValue(ctx, providerSpaceRootLoggerKey+loggerKey("."+subsystem), logger)
+}
+
+func NewSubsystem(ctx context.Context, subsystem string, options ...Option) context.Context {
```

Comment:
Go doc would be especially be helpful here to explain why consumers may want these. 👍

Reply:
Yeah, definitely a thing to do before merging this.